More clearly explain catch binding scope (#20428)
mdn/content: files/en-us/web/javascript/reference/statements/try...catch/index.md

- `tryStatements`
  - : The statements to be executed.
- `catchStatements`
  - : Statement that is executed if an exception is thrown in the `try` block.
- `exceptionVar` {{optional_inline}}

>   - : An optional [identifier or pattern](#the_exception_variable) to hold the caught exception for the associated `catch` block. If the `catch` block does not utilize the exception's value, you can omit the `exceptionVar` and its surrounding parentheses, as `catch {...}`.

Reviewer: ```suggestion
  - : An optional [identifier or pattern](#the_exception_variable) to hold the caught exception for the associated `catch` block. If the `catch` block does not use the exception's value, you can omit the `exceptionVar` and its surrounding parentheses, as `catch {...}`.
```

Reviewer: Maybe just me, but I found "as `catch {...}`" very confusing, it seemed like `{...}` was some special syntax rather than just a depiction of the body of the catch block. Maybe because it was on the same line, or because it read as presented as an alternative to `catch(e)` rather than the whole thing. I don't have an alternative except just omitting the example, which (IMO) would be clearer.

Author: ```suggestion
  - : An optional [identifier or pattern](#the_exception_variable) to hold the caught exception for the associated `catch` block. If the `catch` block does not use the exception's value, you can omit the `exceptionVar` and its surrounding parentheses.
```

```js
try {
  throw new TypeError("oops");
} catch ({ name, message }) {
  console.log(name); // "TypeError"
  console.log(message); // "oops"
}
```

The variables bound by the `catch` clause live in the same scope as the `catch` block, so any variables declared in the `catch` block cannot have the same name as the variables bound by the `catch` clause. (There's [one exception to this rule](/en-US/docs/Web/JavaScript/Reference/Deprecated_and_obsolete_features#statements), but it's a deprecated syntax.)

```js example-bad
try {
  throw new TypeError("oops");
} catch ({ name, message }) {
  var name; // SyntaxError: Identifier 'name' has already been declared
  let message; // SyntaxError: Identifier 'name' has already been declared
}
```

The exception variable is writable. For example, you may want to normalize the exception value to make sure it's an {{jsxref("Error")}} object.

```js
try {
  throw "Oops; this is not an Error object";
} catch (e) {
  if (!(e instanceof Error)) {
    e = new Error(e);
  }
  console.error(e.message);
}
```

> If you don't need the exception value, it could be omitted along with the enclosing parentheses.

Reviewer: ```suggestion
If you don't need the exception value, you can omit it along with the enclosing parentheses.
```

## Examples

### Unconditional catch block

When a `catch` block is used, the `catch` block is executed when any exception is thrown from within the `try` block. For example, when the exception occurs in the following code, control transfers to the `catch` block.

```js
try {
  throw "myException"; // generates an exception
} catch (e) {
  // statements to handle any exceptions
  logMyErrors(e); // pass exception object to error handler
}
```

> The `catch` block specifies an identifier (`e` in the example above) that holds the value of the exception; this value is only available in the {{Glossary("Scope", "scope")}} of the `catch` block.

Reviewer: This seems like it's just repeating what's said above. Especially things like:

> The `catch` block specifies an identifier (`e` in the example above) that holds the value of the exception; this value is only available in the {{Glossary("Scope", "scope")}} of the `catch` block.

looks like substantive documentation, not illustrative example.

Reviewer: In general I wonder a bit about this sequence of three examples (in the old page as well as the new):
- Unconditional catch block
- Conditional catch blocks
- Re-throwing errors

So the first of these just looks like " a catch block", and the second I think "so what, you can use conditional code in a catch block, why wouldn't you be able to". The payoff ISTM is re-throwing errors, which is such a common usage that I wonder if it would be better to rework this to lead with (and in fact only include) that:

```
### Re-throwing errors

A common pattern is to (handle some errors and re-throw the others)
```

...I don't know if that would sit better in "Examples" or "Description". Examples maybe.  I wish again that we had H3s in the ToC.

Author: > so what, you can use conditional code in a catch block, why wouldn't you be able to

The point of this example is not to illustrate a capability, but to mimic a pattern in other languages. For example, in Java you can do this:

```java
try {
  doSomething();
} catch (TypeError e) {
  handle(e);
} catch (RangeError e) {
  handle(e);
}
```

There have been many proposals to implement the same pattern in JavaScript, but others believe that `if...else` within the `catch` block is sufficient.

Author: I understand it's slightly repetitive, but I think it's fine to describe the same concept again in the context of an example.

Reviewer: > The point of this example is not to illustrate a capability, but to mimic a pattern in other languages.

Although I merged this: if the intention here is to contrast with languages which have this feature, then it ought to say that. Many people reading these docs will not have experience of other languages, and although I do, it is rusty enough for me not to understand the point of this example. If real people (i.e. me) don't understand the point of the example, it's a little dismissive to, well, dismiss it.

Author: I will add it to https://github.com/mdn/content/pull/26973, which is enabled by this PR.
